gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnResponseEndpoint.java

```diff
@@ -141,16 +162,40 @@ public void handle(RoutingContext ctx) {
                                         ctx.fail(401);
                                         return;
                                     }
-                                    // save the user and login completed status into the context
-                                    ctx.getDelegate().setUser(user);
-                                    ctx.session().put(ConstantKeys.PASSWORDLESS_AUTH_COMPLETED_KEY, true);
-                                    ctx.session().put(ConstantKeys.WEBAUTHN_CREDENTIAL_ID_CONTEXT_KEY, credentialId);
-                                    ctx.session().put(ConstantKeys.USER_LOGIN_COMPLETED_KEY, true);
-
-                                    // Now redirect back to authorization endpoint.
-                                    final MultiMap queryParams = RequestUtils.getCleanedQueryParams(ctx.request());
-                                    final String returnURL = getReturnUrl(ctx, queryParams);
-                                    ctx.response().putHeader(HttpHeaders.LOCATION, returnURL).end();
+
+                                    Optional<Factor> clientFido2Factor = getClientFido2Factor(client);
+                                    if (isPasswordlessLoggingProcess(ctx) && clientFido2Factor.isPresent()) {
+                                        Optional<EnrolledFactorSecurity> enrolledFido2FactorSecurity = getEnrolledFido2FactorSecurity(authenticatedUser);
+                                        if (enrolledFido2FactorSecurity.isPresent() && enrolledFido2FactorSecurity.get().getValue().equals(credentialId)) {
+                                            //user already has fido2 factor for this credential
+                                            updateSessionAuthAndChallengeStatus(ctx);
+                                            updateSessionLoginCompletedStatus(ctx, user,credentialId);
+                                            redirectToAuthorize(ctx);
+                                        } else {
+                                            //save the fido factor
+                                            saveFidoFactor(authenticatedUser, credentialId, clientFido2Factor.get().getId()).subscribe(
+                                                    () -> {
+                                                        updateSessionAuthAndChallengeStatus(ctx);
+                                                        updateSessionLoginCompletedStatus(ctx, user,credentialId);
+                                                        redirectToAuthorize(ctx);
+                                                    },
+                                                    error -> logger.error("Could not update user profile with FIDO2 factor detail", error)
```

Comment:
We just log the error here, we should also do something on the `RoutingContext` to continue or return an error (`ctx.failure`) otherwise in this case, GW will never respond

Reply:
now returns 401